doc/designs/quic-design/quic-io-arch.md

```diff
@@ -0,0 +1,485 @@
+   for application use of the libssl APIs.
+
+ - In the case of non-blocking applications, it must be possible
+   for an application to do its own polling and make its own event
+   loop.
+
+Requirements
+------------
+
+These requirements are complicated by the fact that traditional use of the
+libssl API allows an application to pass an arbitrary BIO to an SSL object; not
+only that, separate BIOs can be passed for the read and write directions. The
+nature of this BIO can be arbitrary; it could be a socket, or a memory buffer.
+
+Implementation of QUIC will require that the underlying network BIO passed to
+the QUIC implementation be configured to support datagram semantics instead of
+bytestream semantics as has been the case with traditional TLS over TCP.
+
+Implementation of QUIC requires handling of timer events as well as the
+circumstances where a network socket becomes readable or writable. In many cases
+we need to handle these events simultaneously (e.g. wait until a socket becomes
+readable, or a timeout expires, whichever comes first).
+
+Blocking vs. Non-Blocking I/O
+-----------------------------
+
+The above constraints make it effectively a requirement that non-blocking I/O be
+used for the calls to the underlying network BIOs. To illustrate this point, we
+first consider how QUIC might be implemented using blocking I/O internally.
+
+To function correctly and provide blocking semantics at the application level,
+our QUIC implementation must be able to block such that it can respond to any of
+the following events for the underlying network read and write BIOs immediately:
+
+- The underlying network write BIO becomes writeable;
+- The underlying network read BIO becomes readable;
+- A timeout expires.
+
+### Blocking sockets and select(3)
+
+Firstly, consider how this might be accomplished using the Berkeley sockets API.
+Blocking on all three wakeup conditions listed above would require use of an API
+such as select(3) or poll(3), regardless of whether the network socket is
+configured in blocking mode or not.
+
+While in principle APIs such as select(3) can be used with a socket in blocking
+mode, this is not an advisable usage mode. If a socket is in blocking mode,
+calls to send(3) or recv(3) may block for some arbitrary period of time, meaning
+that our QUIC implementation cannot handle incoming data (if we are blocked on
+send), send outgoing data (if we are blocked on receive), or handle timeout
+events.
+
+Though it can be argued that a select(3) call indicating readability or
+writeability should guarantee that a subsequent send(3) or recv(3) call will not
+block, there are several reasons why this is an extremely undesirable solution:
+
+- It is quite likely that there are buggy OSes out there which perform spurious
+  wakeups from select(3).
+
+- The fact that a socket is writeable does not necessarily mean that a datagram
+  of the size we wish to send is writeable, so a send(3) call could block
+  anyway.
```

Comment:
Writes aren't the only possible byproduct of a timeout.

Reply:
> Writes aren't the only possible byproduct of a timeout.

What other byproducts are there - and what is the impact of them being delayed by a small amount?

Reply:
When we get a timeout it could cause a write, but it could also cause connection teardown (idle timeout), meaning application SSL API calls need to return, etc.

I recall seeing research suggesting that the latency of the path to the network is actually quite a relevant factor for optimising QUIC performance. Unfortunately I can't seem to place it currently. It's true that delays on an OS's kernel write queues will probably be short enough that it's not going to be a severe detriment to performance. I am quite concerned however by the prospect of walling ourselves in here and preventing ourselves from tuning this properly in the future.